File: app/src/main/java/de/passwordvault/model/storage/csv/CsvParser.java
package de.passwordvault.model.storage.csv;

import java.util.ArrayList;


/**
 * Class implements a parser which can parse CSV (that was generated through {@link CsvBuilder}).
 *
 * @author  Christian-2003
 * @version 3.0.0
 */
public class CsvParser extends CsvConfiguration {

    /**
     * Attribute stores the parsed CSV columns.
     */
    private final ArrayList<String> columns;

    /**
     * Attribute stores the CSV which shall be parsed.
     */
    private final String csv;

    /**
     * Attribute stores the current index within the parsed CSV.
     */
    private int currentIndex;


    /**
     * Constructor instantiates a new {@link CsvParser}.
     *
     * @param csv                   CSV which shall be parsed.
     * @throws NullPointerException The passed CSV is {@code null}.
     */
    public CsvParser(String csv) throws NullPointerException {
        if (csv == null) {
            throw new NullPointerException("Null is invalid CSV");
        }
        this.csv = csv;
        columns = new ArrayList<>();
        currentIndex = 0;
    }


    /**
     * Method parses the provided CSV. The CSV will be converted into an array list of strings.
     *
     * @return  List of string-contents of the CSV columns.
     */
    public ArrayList<String> parseCsv() {
        columns.clear();
        while (currentIndex < csv.length()) {
            next();
        }
        return columns;
    }


    /**
     * Method parses the next column of the passed CSV and appends the parsed value to
     * {@link #columns}.
     */
    private void next() {
        boolean containsSpecialCharacters = false;
        if (csv.charAt(currentIndex) == STRING_SEPARATOR) {
            containsSpecialCharacters = true;
            currentIndex++;
        }
        StringBuilder column = new StringBuilder();
        for (int i = currentIndex; i < csv.length(); i++) {
            char currentChar = csv.charAt(i);
            switch (currentChar) {
                case STRING_SEPARATOR:
                case COLUMN_DIVIDER:
                case ROW_DIVIDER:
                    if (containsSpecialCharacters && currentChar != STRING_SEPARATOR) {
                        column.append(currentChar);
                    }
                    else {
                        columns.add(column.toString());
                        //Add (i + 2) to index if the current char is string separator, since
                        //character after string separator must always be a column divider:
                        currentIndex = currentChar == STRING_SEPARATOR ? i + 2 : i + 1;
                        return;
                    }
                    break;
                case '\\':
                    if (csv.length() - 1 > i + 1) {
                        char escapedChar = csv.charAt(++i);
                        if (escapedChar == 'n') {
                            column.append(ROW_DIVIDER);
                        }
                        else if (escapedChar == STRING_SEPARATOR) {
                            column.append(STRING_SEPARATOR);
                        }
                        else {
                            column.append(currentChar);
                            i--;
                        }
                    }
                    else {
                        column.append(currentChar);
                    }
                    break;
                default:
                    column.append(currentChar);
            }
        }
        //Parsed the last column in the last row:
        currentIndex = csv.length();
        columns.add(column.toString());
    }

}

File: core/src/main/java/de/christian2003/core/conversion/csv/CsvParser.kt
package de.christian2003.core.conversion.csv


/**
 * Class implements a parser which can parse a row of CSV content.
 *
 * @param config    Config to use when parsing the CSV.
 * @param csv       CSV to parse.
 * @author          Christian-2003
 * @since           3.8.0
 */
class CsvParser(

    /**
     * CSV to parse.
     */
    private val csv: String,

    /**
     * Config to use when parsing the CSV.
     */
    private val config: CsvConfig = CsvConfig()

) {

    /**
     * Attribute stores the current index within the CSV.
     */
    private var currentIndex: Int = 0


    /**
     * Parses the next column of the CSV. The column content will be returned as string (or null if
     * the content of the column is null). If no more columns are available, null is returned.
     *
     * @return  Content of the next column in the CSV.
     */
    fun next(): String? {
        if (currentIndex >= csv.length) {
            return null
        }
        var containsSpecialCharacters = false
        if (csv[currentIndex] == config.stringSeparator) {
            containsSpecialCharacters = true
            currentIndex++
        }
        val column = StringBuilder()
        var i = currentIndex
        while (i < csv.length) {
            when (val char: Char = csv[i]) {
                config.columnDivider, config.rowDivider, config.stringSeparator -> {
                    if (containsSpecialCharacters && char != config.stringSeparator) {
                        column.append(char)
                    }
                    else {
                        //Add (i + 2) to index if the current char is string separator, since char
                        //after string separator must always be a column divider:
                        currentIndex = if (char == config.stringSeparator) { i + 2 } else { i + 1 }
                        return if (column.toString() == "\u0000") {
                            null
                        } else {
                            column.toString()
                        }
                    }
                }
                '\\' -> {
                    if (csv.length - 1 > i + 1) {
                        when (csv[++i]) {
                            'n' -> {
                                column.append(config.rowDivider)
                            }
                            config.stringSeparator -> {
                                column.append(config.stringSeparator)
                            }
                            else -> {
                                column.append(char)
                                i--
                            }
                        }
                    }
                    else {
                        column.append(char)
                    }
                }
                else -> {
                    column.append(char)
                }
            }
            i++
        }

        //Parsed last column in the last row
        currentIndex = csv.length
        return if (column.toString() == "\u0000") {
            null
        } else {
            column.toString()
        }
    }


    /**
     * Returns whether the CSV has a next column to parse.
     *
     * @return  Whether the CSV has a next column to parse.
     */
    fun hasNext(): Boolean {
        return currentIndex < csv.length
    }

}
